packages/region/src/services/region-module.ts

```diff
@@ -0,0 +1,230 @@
+import {
+  Context,
+  CreateRegionDTO,
+  DAL,
+  InternalModuleDeclaration,
+  IRegionModuleService,
+  ModuleJoinerConfig,
+  ModulesSdkTypes,
+  RegionCountryDTO,
+  RegionCurrencyDTO,
+  RegionDTO,
+  UpdateRegionDTO,
+} from "@medusajs/types"
+import {
+  InjectManager,
+  InjectTransactionManager,
+  MedusaContext,
+  MedusaError,
+  ModulesSdkUtils,
+  promiseAll,
+} from "@medusajs/utils"
+
+import { Country, Currency, Region } from "@models"
+
+import { DefaultsUtils } from "@medusajs/utils"
+import { CreateCountryDTO, CreateCurrencyDTO } from "@types"
+import { entityNameToLinkableKeysMap, joinerConfig } from "../joiner-config"
+
+type InjectedDependencies = {
+  baseRepository: DAL.RepositoryService
+  regionService: ModulesSdkTypes.InternalModuleService<any>
+  countryService: ModulesSdkTypes.InternalModuleService<any>
+  currencyService: ModulesSdkTypes.InternalModuleService<any>
+}
+
+const generateMethodForModels = [Country, Currency]
+
+export default class RegionModuleService<
+    TRegion extends Region = Region,
+    TCountry extends Country = Country,
+    TCurrency extends Currency = Currency
+  >
+  extends ModulesSdkUtils.abstractModuleServiceFactory<
+    InjectedDependencies,
+    RegionDTO,
+    {
+      Country: {
+        dto: RegionCountryDTO
+        singular: "Country"
+        plural: "Countries"
+      }
+      Currency: {
+        dto: RegionCurrencyDTO
+        singular: "Currency"
+        plural: "Currencies"
+      }
+    }
+  >(Region, generateMethodForModels, entityNameToLinkableKeysMap)
+  implements IRegionModuleService
+{
+  protected baseRepository_: DAL.RepositoryService
+  protected readonly regionService_: ModulesSdkTypes.InternalModuleService<TRegion>
+  protected readonly countryService_: ModulesSdkTypes.InternalModuleService<TCountry>
+  protected readonly currencyService_: ModulesSdkTypes.InternalModuleService<TCurrency>
+
+  constructor(
+    {
+      baseRepository,
+      regionService,
+      countryService,
+      currencyService,
+    }: InjectedDependencies,
+    protected readonly moduleDeclaration: InternalModuleDeclaration
+  ) {
+    // @ts-ignore
+    super(...arguments)
+    this.baseRepository_ = baseRepository
+    this.regionService_ = regionService
+    this.countryService_ = countryService
+    this.currencyService_ = currencyService
+  }
+
+  __hooks = {
+    onApplicationStart: async () => await this.createCountriesAndCurrencies(),
+  }
```

Comment:
thought(non-blocking): this looks super ugly - the `__` makes it feel like it's something you really shouldn't use. Gives me these vibes: `React.__SECRET_DOM_DO_NOT_USE_OR_YOU_WILL_BE_FIRED`.

Let's revisit at some point to make the API for this better. Same goes for joinerConfig